blockstore/apps/bundles/data/store.py

```diff
@@ -0,0 +1,222 @@
+    public = attr.ib(type=bool)
+    size = attr.ib(type=int)
+    hash = attr.ib()  # Actual Hash object, not a string
+
+@attr.s(frozen=True)
+class Link():
+    alias = attr.ib(type=str)
+    target = attr.ib(type=BundleVersionKey)
+    uses = attr.ib(type=list)
+
+
+@attr.s()
+class DependencyMapping():
+    # Map of strings (no /'s allowed) to BundleVersionKey objects
+    links = attr.ib(type=dict)
+
+    def add_link(self, link):
+        """"""
+        pass
+
+    def all(self):
+        """Return full list of all BundleVersionKeys"""
+        pass
+
+
+@attr.s()
+class BundleVersionSummary():
+    """
+    All the read-only attributes of a BundleVersion.
+
+    Common logic like dependency calculation can happen here.
+    """
+    bundle_uuid = attr.ib()
+
+    # Map of paths to FileInfo objects
+    files = attr.ib(type=dict)
+
+    # All our various links and their extended dependencies
+    dependencies = attr.ib(type=DependencyMapping)
+
+class BundleJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, BundleVersionKey):
+            return str(obj)
+        elif isinstance(obj, FileInfo):
+            return [
+                obj.path,
+                obj.public,
+                obj.size,
+                obj.hash.hexdigest(),
+            ]
+        elif isinstance(obj, pathlib.Path):
+            return str(obj)
+        elif isinstance(obj, BundleVersionSummary):
+            return {
+                'bundle_uuid': str(obj.bundle_uuid),
+                'files': obj.files,
+                'dependencies': obj.dependencies,
+            }
+        return json.JSONEncoder.default(self, obj)
+
+
+class BundleDataStore():
+    """
+    This is the interface for storing the actual files that make up a Bundle.
+
+    The responsibilities of this class are:
+    * Create a new Snapshot from a set of files.
+    * Generate URLs for resources within a Bundle.
+    * Validate Links.
+
+    # Important note about Versions/Snapshots:
+
+    Numbered versions don't exist at this level -- versioning here is entirely
+    based on content, so basically some kind of hash. BundleVersion is what
+    manages numbered versioning and points to an entry in BundleDataStore. To
+    make that distinction clearer, we'll say that BundleDataStore creates
+    Snapshots, which BundleVersion will point to.
+
+    This also helps us to mitigate the concurrency issue. Creating a snapshot
+    might take a long time (say to upload all the assets). If we were assigning
+    a version_num at the start of that process, we might run into a race
+    condition with a process that starts a few seconds later, because they'll
+    both see the database reflecting the same state. But if BundleStore creates
+    only hash-versioned Snapshots, then it can do all that work and the only
+    time version_num is read and incremented for a new entry is when
+    BundleDataStore emits a signal that BundleVersion catches.
```

Comment:
Actually, I guess since the only thing that needs version information in the BundleStore is the summary JSON, we still wouldn't have much in the way of concurrency issues because we could delay until then. But that would involve it calling up the abstraction stack a layer, which still feels wrong.